[TRTLLM-5670][feat] Support top log probabilities

stnie opened the pull request:
# [TRTLLM-5670][feat] Support top log probabilities

- Added a new parameter `topLogProbs` to the `OutputConfig` class, allowing users to specify the number of top log probabilities to return.
- Updated related methods and bindings in the executor and API layers to accommodate this feature.
- updated code in Torch sampler to provide the functionality
- added a new Dict to LogProbStorage containing the top logprobs
- included validation checks for top log probabilities in the sampling parameters
- added tests to ensure correct functionality.

**Change:** Changed API to conform better with current API.
 - removed top_logprobs parameter and used logprobs instead
 - concatenated top_logprobs and logprobs in output instead of adding top_logprobs field
 - Note: Need to update Performance tests for this new change.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Return top token log-probabilities with each generated token.
  - New controls: SamplingParams.top_logprobs and LLM max_top_logprobs.
  - Python API/bindings accept, serialize, and expose top_logprobs across requests and results.

- Documentation
  - quickstart_advanced adds --top_logprobs and --max_top_logprobs flags, prints top log-probs, and auto-enables mixed sampling when applicable.

- Tests
  - Added unit and integration tests covering top_logprobs validation, serialization, and runtime behavior (including torch sampler paths).
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

Allow the user to specify with the top_logprobs parameter if he wants to have the logprobs of the top <top_logprobs> tokens at each step in addition to the logprobs of the sampled tokens. 

Currently only functionality was tested. 

## Performance Tests
I ran performance checks of main and this branch for greedy and top-k sampling with logprobs=false/true to see how the changes impact workflows, where top-logprobs are not enabled.

Number of requests:             100
Number of concurrent requests:  55.9995
Average Input Length (tokens):  40.0000
Average Output Length (tokens): 2009.0000
TP Size:                1
PP Size:                1
EP Size:                None
Max Runtime Batch Size: 64
Max Runtime Tokens:     131072
Scheduling Policy:      GUARANTEED_NO_EVICT
KV Memory Percentage:   90.00%
Issue Rate (req/sec):   3.9227E+17


### This branch (Need update since latest changes)

Test Configuration | Request Throughput (req/sec) | Total Output Throughput (tokens/sec) | Total Token Throughput (tokens/sec) | Total Latency (ms) | Average Request Latency (ms)
--- | --- | --- | --- | --- | ---
Greedy Sampling (w/ logprobs) | 1.8135 | 3643.3317 | 3715.8719 | 55141.8369 | 30763.6410
Greedy Sampling (w/o logprobs) | 3.1688 | 6366.1379 | 6492.8903 | 31557.5947 | 18070.6263
Top-k Only Sampling (w/logprobs) | 1.7962 | 3608.5418 | 3680.3893 | 55673.4576 | 31001.0933
Top-k Only Sampling (w/o logprobs) | 2.7684 | 5561.8000 | 5672.5377 | 36121.3996 | 20532.6132

### main branch

Test Configuration | Request Throughput (req/sec) | Total Output Throughput (tokens/sec) | Total Token Throughput (tokens/sec) | Total Latency (ms) | Average Request Latency (ms)
--- | --- | --- | --- | --- | ---
Greedy Sampling (w/ logprobs) | 1.8263 | 3669.0145 | 3742.0661 | 54755.8476 | 30663.0136
Greedy Sampling (w/o logprobs | 3.1843 | 6397.2784 | 6524.6508 | 31403.9797 | 17990.4672
Top-k Only Sampling (w/ logprobs) | 1.7546 | 3524.9074 | 3595.0897 | 56994.4056 | 31702.4851
Top-k Only Sampling (w/o logprobs) | 2.7831 | 5591.2461 | 5702.5701 | 35931.1673 | 20441.4355



## Test Coverage

new test: test_return_top_k_logprobs.py 
tests if the shape of the results is correct. Additionally ensures that some failure cases throw the correct error

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

stnie committed:
[TRTLLM-5670][feat] Enhance OutputConfig to support top log probabilities

- Added a new parameter `topLogProbs` to the `OutputConfig` class, allowing users to specify the number of top log probabilities to return.
- Updated related methods and bindings in the executor and API layers to accommodate this feature.
- updated code in Torch sampler to provide the functionality
- added a new Dict to LogProbStorage containing the top logprobs
- included validation checks for top log probabilities in the sampling parameters
- added tests to ensure correct functionality.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #16678](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16678/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #12519](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12519/) completed with status: 'FAILURE'

stnie commented:
/bot run

stnie commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -376,17 +378,19 @@ class Args:
 
         self.NEW_TOKENS_SHAPE = (self.max_tokens, self.max_num_sequences,
-                                 self.MAX_BEAM_WIDTH)
+                                 self.MAX_BEAM_WIDTH, 1 + self.max_top_logprobs)
```
Because we need the information regarding the top tokens to provide a matching between top-tokens and top-logprobs. This shape is split into two, when creating the SampleState

stnie committed:
[TRTLLM-5670][chore] Update OutputConfig and related components for optional top log probabilities

- Changed `topLogProbs` parameter in `OutputConfig` to be optional, allowing for more flexible configurations.
- Updated related methods and bindings in the executor and API layers to handle the new optional type.
- Enhanced validation checks in sampling parameters to ensure correct usage of `top_logprobs`.
- Adjusted tests to accommodate the changes in parameter handling and validation.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

stnie commented:
/bot run

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16804](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16804/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #16805](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16805/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #12615](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12615/) completed with status: 'FAILURE'

stnie committed:
[TRTLLM-5670][chore] Refine several small bugs / descriptions

- Updated `OutputConfig` to use `std::optional<SizeType32>` for the `top_logprobs` parameter
- Adjusted docstring in `sampling_params.py` to correctly state the default of `top_logprobs` as None.
- Modified `quickstart_advanced.py` to ensure correct handling of `top_k` conditions.
- Clarified documentation in `sampler.py` regarding indices after applying top-k and top-p masks.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

stnie commented on cpp/tensorrt_llm/nanobind/executor/request.cpp:
```diff
@@ -216,24 +216,26 @@ void initRequestBindings(nb::module_& m)
                     return_generation_logits.value_or(false), exclude_input_from_output.value_or(false),
                     return_encoder_output.value_or(false), return_perf_metrics.value_or(false),
-                    additional_model_outputs);
+                    additional_model_outputs, top_logprobs.value_or(0));
```
done

stnie commented on cpp/include/tensorrt_llm/executor/executor.h:
```diff
@@ -215,7 +215,8 @@ class OutputConfig
         bool returnPerfMetrics = false,
-        std::optional<std::vector<AdditionalModelOutput>> additionalModelOutputs = std::nullopt);
+        std::optional<std::vector<AdditionalModelOutput>> additionalModelOutputs = std::nullopt,
+        SizeType32 topLogProbs = 0);
```
done

stnie commented on cpp/tensorrt_llm/nanobind/executor/request.cpp:
```diff
@@ -216,24 +216,26 @@ void initRequestBindings(nb::module_& m)
         .def_rw("return_encoder_output", &tle::OutputConfig::returnEncoderOutput)
         .def_rw("return_perf_metrics", &tle::OutputConfig::returnPerfMetrics)
         .def_rw("additional_model_outputs", &tle::OutputConfig::additionalModelOutputs)
+        .def_rw("top_logprobs", &tle::OutputConfig::topLogProbs)
```
I adjusted these two tests

tensorrt-cicd commented:
[PR_Github #16817](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16817/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #12624](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12624/) completed with status: 'FAILURE'

stnie committed:
[TRTLLM-5670][chore] Update autodeploy code to adjust to the top_logprobs changes

- Correctly set max_top_logprobs parameter in create_autodeploy_executor.
- Updated DemoEngine to correctly unpack top_k and greedy_search

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16851](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16851/) [ run ] triggered by Bot

stnie commented:
/bot run --disable-fail-fast

stnie committed:
[TRTLLM-5670][chore] Add top_logprobs parameter to CompletionOutput docstring and updated yaml files for api stability testing

- Updated the test cases in `test_llm_api.py`, `completion_output.yaml`, `llm.yaml`, and `sampling_params.yaml` to include the new `top_logprobs` parameter and its associated properties.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17516](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17516/) [ run ] triggered by Bot

venkywonka commented:
Naive Q, won't `top_logprobs` make `logprobs` redundant?

stnie commented:
@venkywonka
Note that In the current state `logprobs` is used only as a boolean in the PyTorch Flow. 
I added the `top_logprobs `to have a way to differentiate the cases, where only the `logprobs `of the sampled tokens should be returned `(logprobs=True , top_logprobs=None`) vs the case where the `logprobs `of both: the sampled tokens and the top-1 token should be returned (`logprobs=True, top_logprobs = 1`). Having only one parameter (e.g.` logprobs = 1`)  would be ambiguous.

stnie committed:
[TRTLLM-5670][chore] Update top logprobs to reflect changes in sampler

- Updated `TorchStore` and `TorchSampler` to include `max_top_logprobs` parameter for managing top log probabilities.
- Modified methods to handle top log probabilities during sampling and logging.
- Added new Tensors to `TorchStore` containing data for top logprobs
- Adjusted tests to validate the correct functionality of top log probabilities in various scenarios.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #18068](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18068/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13543](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13543/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/13543/test-results/rerun_report.html) for details.

stnie committed:
[TRTLLM-5670][chore] Change top logprobs API to conform more with current LLM-API

- Removed the top_logprobs parameter from SamplingParams and related classes to streamline log probability management.
- Change _get_output_config to explicitly pass logprobs as top-logprobs to OutputConfig object.
- Updated setup_llm function to conditionally set logprobs based on input arguments.
- Adjusted CompletionOutput and related tests to reflect the removal of top_logprobs.
- Enhanced validation checks to ensure correct usage of logprobs in sampling scenarios.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #18368](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18368/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13776](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13776/) completed with status: 'FAILURE'

stnie committed:
[TRTLLM-5670][chore] Update logprobs validation for greedy sampling

- Adjusted the validation logic in SamplingParams to ensure that logprobs must not exceed 1 when greedy sampling is enabled.
- Updated error message for clarity regarding the constraints on logprobs in greedy sampling scenarios.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #18609](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18609/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13970](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13970/) completed with status: 'FAILURE'

stnie committed:
[TRTLLM-5670][chore] Update Greedy sampling to also support top logprobs of 1

- Modified the return statement in `greedy_search_sampling_batch` to include `next_tokens` as an unsqueezed tensor, ensuring compatibility with downstream processing.
- Added `max_top_logprobs` parameter to the configuration in the test file to ensure correct execution.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #18738](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18738/) [ run ] triggered by Bot

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #18767](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18767/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18763](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18763/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14065/" target="_blank">LLM/main/L0_MergeRequest_PR #14065</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14065//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

tensorrt-cicd commented:
[PR_Github #18767](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18767/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14069](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14069/) completed with status: 'FAILURE'

stnie committed:
[TRTLLM-5670][chore] Adjust top logprobs handling in tests

- Updated test_embedding_bias_with_torch_sampler_strategies cases to correctly handle top logprobs of 1 in the parameterization
- Enhanced validation logic in test_return_top_k_logprobs to accommodate the new conditions for valid setups in sampling tests (greedy sampling).

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>
